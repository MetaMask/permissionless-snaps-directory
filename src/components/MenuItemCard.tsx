import { Text, HStack, MenuItem } from '@chakra-ui/react';
import type { FunctionComponent } from 'react';

import { Card } from './Card';

export type MenuCardProps = {
  icon: JSX.Element;
  label: string;
  textColor?: 'default' | 'red' | 'blue';
  onClick?: () => void;
};

export const MenuItemCard: FunctionComponent<MenuCardProps> = ({
  icon,
  label,
  textColor = 'default',
  onClick,
}) => {
  return (
    <MenuItem w="95%" onClick={onClick} data-testid="menu-item-card">
      <Card padding="1">
        <HStack spacing="2">
          <Text variant={textColor}>{icon}</Text>
          <Text variant={textColor}>{label}</Text>
        </HStack>
      </Card>
    </MenuItem>
  );
};
