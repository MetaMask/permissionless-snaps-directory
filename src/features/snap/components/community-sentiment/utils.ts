import { t } from '@lingui/macro';

import { SentimentType } from './types';

export const getColorForSentiment = (type: SentimentType) => {
  switch (type) {
    case SentimentType.InsufficientReview:
      return t`warning`;
    case SentimentType.Endorsed:
      return t`success`;
    case SentimentType.InReview:
      return t`warning`;
    case SentimentType.Reported:
      return t`error`;
    default:
      return t`success`;
  }
};

export const getSentimentTypeFromResult = (result: number) => {
  switch (result) {
    case 1:
      return SentimentType.Endorsed;
    case 2:
      return SentimentType.InReview;
    case 3:
      return SentimentType.Reported;
    default:
      return SentimentType.InsufficientReview;
  }
};
