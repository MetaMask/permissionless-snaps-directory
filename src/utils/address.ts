import type { Hex } from '@metamask/utils';
import { getAddress } from 'viem';

export const parseAddress = (address: Hex): Hex | null => {
  try {
    return getAddress(address);
  } catch {
    return null;
  }
};
