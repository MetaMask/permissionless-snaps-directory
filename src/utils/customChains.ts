import type { Address } from 'viem';

const lineaMainnet = {
  id: 59_144,
  name: 'Linea Mainnet',
  network: 'linea-mainnet',
  nativeCurrency: { name: 'Linea Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    infura: {
      http: ['https://linea-mainnet.infura.io/v3'],
      webSocket: ['wss://linea-mainnet.infura.io/ws/v3'],
    },
    default: {
      http: ['https://rpc.linea.build'],
      webSocket: ['wss://rpc.linea.build'],
    },
    public: {
      http: ['https://rpc.linea.build'],
      webSocket: ['wss://rpc.linea.build'],
    },
  },
  blockExplorers: {
    default: {
      name: 'Etherscan',
      url: 'https://lineascan.build/',
    },
  },
  contracts: {
    multicall3: {
      address: '0xca11bde05977b3631167028862be2a173976ca11' as Address,
      blockCreated: 0,
    },
  },
  testnet: false,
};

const lineaTestnet = {
  id: 59140,
  name: 'Linea Goerli Testnet',
  network: 'linea-testnet',
  nativeCurrency: {
    name: 'Linea Ether',
    symbol: 'ETH',
    decimals: 18,
  },
  rpcUrls: {
    infura: {
      http: ['https://linea-goerli.infura.io/v3'],
      webSocket: ['wss://linea-goerli.infura.io/v3'],
    },
    default: {
      http: ['https://rpc.goerli.linea.build'],
      webSocket: ['wss://rpc.goerli.linea.build'],
    },
    public: {
      http: ['https://rpc.goerli.linea.build'],
      webSocket: ['wss://rpc.goerli.linea.build'],
    },
  },
  blockExplorers: {
    default: {
      name: 'Etherscan',
      url: 'https://goerli.lineascan.build',
    },
  },
  contracts: {
    multicall3: {
      address: '0xca11bde05977b3631167028862be2a173976ca11' as Address,
      blockCreated: 498623,
    },
  },
  testnet: true,
};

export { lineaMainnet, lineaTestnet };
