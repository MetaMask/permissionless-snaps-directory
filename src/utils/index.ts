export * from './errors';
export * from './links';
export * from './snaps';
export * from './string';
export * from './address';
export * from './signMessage';
