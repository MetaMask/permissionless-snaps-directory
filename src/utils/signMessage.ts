import type { SignTypedDataArgs } from '@wagmi/core';
import { isAddress } from 'viem';

/**
 * Generate the message containing TrustCredential to be signed.
 *
 * @param issueAddress - The issuer address to be included in the message.
 * @param subjectAddress - The subject address to be included in the message.
 * @param chainId - The chainId to be included in the message.
 * @param isAdd - The boolean to flag whether is for add or remove.
 * @returns The message to be signed.
 */
export function generateAccountTrustMsg(
  issueAddress: string,
  subjectAddress: string,
  chainId: number,
  isAdd: boolean,
): SignTypedDataArgs {
  if (!isAddress(issueAddress)) {
    throw new Error('Invalid issueAddress');
  }

  if (!isAddress(subjectAddress)) {
    throw new Error('Invalid subjectAddress');
  }

  const domain = {
    name: 'EIP712Example',
    version: '1',
    chainId,
  };

  const types = {
    Trustworthiness: [
      {
        name: 'scope',
        type: 'string',
      },
      {
        name: 'level',
        type: 'uint8',
      },
      {
        name: 'reason',
        type: 'string[]',
      },
    ],
    CredentialSubject: [
      {
        name: 'id',
        type: 'string',
      },
      {
        name: 'trustworthiness',
        type: 'Trustworthiness[]',
      },
    ],

    TrustCredential: [
      {
        name: 'type',
        type: 'string[]',
      },
      {
        name: 'issuer',
        type: 'string',
      },
      {
        name: 'credentialSubject',
        type: 'CredentialSubject',
      },
    ],
  };

  const message = {
    type: ['TrustCredential'],
    issuer: `did:pkh:eip155:${chainId}:${issueAddress}`,
    credentialSubject: {
      id: `did:pkh:eip155:${chainId}:${subjectAddress}`,
      trustworthiness: [
        {
          scope: 'Honesty',
          level: isAdd ? 1 : 0,
          reason: [],
        },
      ],
    },
  };

  return { domain, message, primaryType: 'TrustCredential', types };
}
