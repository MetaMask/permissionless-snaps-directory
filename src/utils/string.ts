/**
 * Trim the text by the given length.
 *
 * @param content - The raw text content.
 * @param length - The length to render.
 * @returns The trimed text.
 */
export function trimText(content: string, length = 100) {
  if (!content) {
    return content;
  }

  const _content = content.trim();

  const contentArr = [..._content];
  if (contentArr.length <= length) {
    return _content;
  }

  // Handle sepcial characters
  return `${contentArr.slice(0, length).join('')}...`;
}

/**
 * Trim the text by the heading length and tailing length.
 *
 * @param content - The raw text content.
 * @param head - The heading length to render.
 * @param tail - The tailing length to render.
 * @returns The trimed text.
 */
export function trimTextByHeadTail(
  content: string,
  head: number,
  tail: number,
) {
  if (!content) {
    return content;
  }

  const _content = content.trim();

  // Handle sepcial characters
  const contentArr = [..._content];
  if (contentArr.length <= head + tail) {
    return _content;
  }

  return `${contentArr.slice(0, head).join('')}...${contentArr
    .slice(-tail)
    .join('')}`;
}

/**
 * Renders the short address of a user.
 *
 * @param address - The address to render.
 * @returns The rendered short address.
 */
export function trimAddress(address: string) {
  return trimTextByHeadTail(address, 7, 5);
}
