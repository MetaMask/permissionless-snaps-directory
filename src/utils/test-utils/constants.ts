export const VALID_ACCOUNT_1 = '0x6B24aE0ABbeb67058D07b891aF415f288eA57Cc7';
export const VALID_ACCOUNT_2 = '0x7B17936E2a44C1fFef392d13e490fc8426d3329e';

export const INVALID_ADDRESS = '0x123123123';

export const SNAP_SHASUM_1 = 'ALj0YMDz4JNchPJfreWErQx/GOh4FMWWy1o1rCq7sQQ=';
export const SNAP_SHASUM_2 = 'CaO1hCMftzS8NR608xuMe/kjXG7k+YtG6eq1Q6jAx7Q=';
