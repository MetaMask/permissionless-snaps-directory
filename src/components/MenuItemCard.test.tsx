import { Menu, MenuList } from '@chakra-ui/react';
import { fireEvent, screen } from '@testing-library/react';

import { MenuItemCard } from './MenuItemCard';
import { render } from '../utils/test-utils';

describe('MenuItemCard', () => {
  describe('when the text color is default', () => {
    const mockOnClick = jest.fn();

    beforeEach(() => {
      render(
        <Menu defaultIsOpen={true}>
          <MenuList>
            <MenuItemCard
              icon={<span>Icon</span>}
              label="Label"
              textColor="default"
              onClick={mockOnClick}
            />
          </MenuList>
        </Menu>,
      );
    });

    it('renders the icon and label', () => {
      const icon = screen.getByText('Icon');
      const label = screen.getByText('Label');
      expect(icon).toBeInTheDocument();
      expect(label).toBeInTheDocument();
    });

    it('calls the onClick function when clicked', () => {
      const menuItem = screen.getByRole('menuitem');
      fireEvent.click(menuItem);

      expect(mockOnClick).toHaveBeenCalled();
    });
  });

  describe('when the textColor is red', () => {
    beforeEach(() => {
      render(
        <Menu defaultIsOpen={true}>
          <MenuList>
            <MenuItemCard
              icon={<span>Icon</span>}
              label="Label"
              textColor="red"
            />
          </MenuList>
        </Menu>,
      );
    });

    it('renders the icon and label', () => {
      const icon = screen.getByText('Icon');
      const label = screen.getByText('Label');
      expect(icon).toBeInTheDocument();
      expect(label).toBeInTheDocument();
    });
  });

  describe('when the onClick function is not provided', () => {
    beforeEach(() => {
      render(
        <Menu defaultIsOpen={true}>
          <MenuList>
            <MenuItemCard icon={<span>Icon</span>} label="Label" />
          </MenuList>
        </Menu>,
      );
    });

    it('does not call the onClick function when clicked', () => {
      const menuItem = screen.getByRole('menuitem');
      expect(menuItem).not.toHaveAttribute('onClick');
    });
  });

  it('match the snapshot', () => {
    const { asFragment } = render(
      <Menu defaultIsOpen={true}>
        <MenuList>
          <MenuItemCard icon={<span>Icon</span>} label="Label" />
        </MenuList>
      </Menu>,
    );
    expect(asFragment()).toMatchInlineSnapshot();
  });
});
