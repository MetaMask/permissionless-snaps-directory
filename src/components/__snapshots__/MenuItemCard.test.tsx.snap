// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItemCard match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-r6z5ec"
    style="visibility: visible; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
  >
    <div
      aria-orientation="vertical"
      class="chakra-menu__menu-list css-1kfu8nn"
      id="menu-list-:rd:"
      role="menu"
      style="transform-origin: var(--popper-transform-origin); visibility: visible; opacity: 1; transform: none;"
      tabindex="-1"
    >
      <button
        class="chakra-menu__menuitem css-1qilg0h"
        data-index="0"
        id="menu-list-:rd:-menuitem-:re:"
        role="menuitem"
        tabindex="-1"
        type="button"
      >
        <div
          class="css-ive15a"
        >
          <div
            class="chakra-stack css-1uodvt1"
          >
            <p
              class="chakra-text css-0"
            >
              <span>
                Icon
              </span>
            </p>
            <p
              class="chakra-text css-0"
            >
              Label
            </p>
          </div>
        </div>
      </button>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
